pex/pip.py

```diff
@@ -429,20 +431,48 @@ def _extract_resolve_data(url):
     def analyze(self, line):
         # type: (str) -> _LogAnalyzer.Continue[None]
 
+        # The log sequence for processing a resolved requirement is as follows (log lines irrelevant
+        # to our purposes omitted):
+        #
+        # 1.) "... Found link <url1> ..."
+        # ...
+        # 1.) "... Found link <urlN> ..."
+        # 2.) "... Added <requirement pin> from <url> ... to build tracker ..."
+        # 3.) Lines related to extracting metadata from <requirement pin>'s artifact
+        # 4.) "... Removed <requirement pin> from <url> ... from build tracker ..."
+        #
+        # The lines in section 3 can contain this same pattern of lines if the metadata extraction
+        # proceeds via PEP-517 which recursively uses Pip to resolve build dependencies. We want to
+        # ignore this recursion since a lock should only contain install requirements and not build
+        # requirements (If a build proceeds differently tomorrow than today then we don't care as
+        # long as the final built artifact hashes the same. In other words, we completely rely on a
+        # cryptographic fingerprint for reproducibility and security guarantees from a lock).
```

Comment:
Oh, I'm not that familiar with Python package builds. My experience from building binaries (such as for C applications, for instance) is that even when the source input is the same, the binary output usually differ (due to embedded timestamps if nothing else). Is this not the case for Python applications/libraries? (as that would break the fingerprint)

Reply:
Its should not be the case for Pex. There are some standards out in the world now that help even beyond python and like this one which we also use:
https://github.com/pantsbuild/pex/blob/7cfb36d9ad094b5dcfe0ae443f3470e244ea67ed/pex/tools/commands/repository.py#L248-L254
https://github.com/pantsbuild/pex/blob/56455c4f307c024b89738d5ad59468600a8c4c2f/pex/bin/pex.py#L724-L728

We have semi-extensive tests for this, but I'm pretty sure we don't set `SOURCE_DATE_EPOCH` for our `pip download ...` and `pip wheel ...` commands, both of which might build an sdist into a wheel.

Reply:
Ah, right. We do not in fact set `SOURCE_DATE_EPOCH` when running `pip download ...` or `pip wheel ...`; so the resulting wheels do have non-reproducible timestamps for their zip entries, but since PEX files don;t use wheels, but _installed_ wheels, the non-reproducible timestamps in the wheels we installed from evaporate. If and when Pex starts doing all the gymnastics to support VCS requirements and local project requirements in locks, then that code will need to handle this. This code, though, just uses fingerprints of downloaded sdists or wheels. The building going on here is just to extract metadata for the resolve; i.e.: Find out the install_requires dependencies of sdists to continue the recursive resolve process. The only thing that needs to be reproducible for that are the Requires-Dist metadata lines.